### ISSUE TITLE
Error: pandora.errors.InvalidAuthToken: Invalid Auth Token

### ISSUE THREAD

**Issue by ksaye** [mcrute/pydora#69]
**What happened**:
As of yesterday, I am getting the error:
`pandora.errors.InvalidAuthToken: Invalid Auth Token`

**What you expected to happen**:
As has worked for years, I expect it to login

**How to reproduce it (as minimally and precisely as possible)**:
1. on Python 3.8 (on Linux or Windows), install the pydora
2. using a new or old pandora account (not pandora one) run the following python script, adding the partner key and your username and password:

        from pandora.clientbuilder import SettingsDictBuilder
        from pandora import client
        
        userName="k******g"
        passWord="M******1"
        
        client = SettingsDictBuilder({
            "DECRYPTION_KEY": "R****B#",    
            "ENCRYPTION_KEY": "6****D",    
            "PARTNER_USER": "android", 
            "PARTNER_PASSWORD": "A******7",    
            "DEVICE": "android-generic",
        }).build()
        
        client.login(userName, passWord)
You should get the following error:

    C:/Python38-64/python.exe d:/Git/PandoraPlayer/getMusic.py
    Traceback (most recent call last):
      File "d:/Git/PandoraPlayer/getMusic.py", line 19, in <module>
        client.login(userName, passWord)
      File "C:\Python38-64\lib\site-packages\pandora\client.py", line 64, in login
        return self._authenticate()
      File "C:\Python38-64\lib\site-packages\pandora\client.py", line 70, in _authenticate
        user = self.transport(
      File "C:\Python38-64\lib\site-packages\pandora\transport.py", line 51, in function
        return func(*args, **kwargs)
      File "C:\Python38-64\lib\site-packages\pandora\transport.py", line 243, in __call__
        return self._parse_response(result)
      File "C:\Python38-64\lib\site-packages\pandora\transport.py", line 232, in _parse_response
        raise PandoraException.from_code(result["code"], result["message"])
    pandora.errors.InvalidAuthToken: Invalid Auth Token

**Anything else we need to know?**:
I have tried this with the iphone partner id, with 2 known good pandora login accounts and with a new pandora login account, all the same result.

Yesterday it was working fine so, I am confused.

**Environment**:
- Pydora version: 2.0.1
- Python version: 3.8.5 - x64
- OS: Windows 10
- Shell (if applicable):
- Others:

**Comment by ksaye** [comment]
I just verified the issue also happens on pydora 2.1.1

**Comment by ksaye** [comment]
It seems that it works now, so I am guessing there might have been an error with Pandora?

**Comment by ksaye** [comment]
@mcrute , I am seeing this exact issue again.  Any suggestion?

**Comment by zonywhoop** [comment]
Seeing this same issue as well on my side using Mopidy-Pandora client

* Python 3.8.10
* Ubuntu 20.04.3
* pydora 2.1.1

Using the android partner passwords in the config.

```
Dec 22 22:25:26 dabox mopidy[376109]: ERROR    [PandoraBackend-7] pykka Unhandled exception in PandoraBackend (urn:uuid:333f3d23-36d5-45b8-8696-7150d87cf649):
Dec 22 22:25:26 dabox mopidy[376109]: Traceback (most recent call last):
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/lib/python3/dist-packages/pykka/_actor.py", line 186, in _actor_loop
Dec 22 22:25:26 dabox mopidy[376109]:     self.on_start()
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/mopidy_pandora/backend.py", line 48, in on_start
Dec 22 22:25:26 dabox mopidy[376109]:     self.api.login(self.config["username"], self.config["password"])
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/pandora/client.py", line 64, in login
Dec 22 22:25:26 dabox mopidy[376109]:     return self._authenticate()
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/pandora/client.py", line 70, in _authenticate
Dec 22 22:25:26 dabox mopidy[376109]:     user = self.transport(
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/pandora/transport.py", line 51, in function
Dec 22 22:25:26 dabox mopidy[376109]:     return func(*args, **kwargs)
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/pandora/transport.py", line 243, in __call__
Dec 22 22:25:26 dabox mopidy[376109]:     return self._parse_response(result)
Dec 22 22:25:26 dabox mopidy[376109]:   File "/usr/local/lib/python3.8/dist-packages/pandora/transport.py", line 232, in _parse_response
Dec 22 22:25:26 dabox mopidy[376109]:     raise PandoraException.from_code(result["code"], result["message"])
Dec 22 22:25:26 dabox mopidy[376109]: pandora.errors.InvalidAuthToken: Invalid Auth Token
Dec 22 22:25:26 dabox mopidy[376109]: ERROR    [MainThread] mopidy.commands Actor died: PandoraBackend (urn:uuid:333f3d23-36d5-45b8-8696-7150d87cf649) stopped before handling the message
```

**Comment by mcrute** [comment]
TL;DR: Use the AIR client ID, this is probably not going to be fixable any time soon.

I can reproduce this as well. In-fact the only credential set that seems to work for me anymore is the one from the AIR client. I guess the best solution for now is to use that credential and disregard the rest. Note that this will not work with free accounts. At this point there is no path forward for free accounts.

I don't know what's going on but I have a guess. Pandora is powering their web player with a newer REST API that's a lot more modern and straightforward than the legacy API that Pydora uses. At the same time it looks like they either have or are discontinuing Pandora One in favor of Pandora Plus and Pandora Premium. So my guess then is that they're turning down their legacy API in favor of the new one and they're doing it key by key.

Looking at their API, it's JSON/REST/GraphQL once you have an auth token. Getting an auth token isn't easy though because they're using something called [HUMAN Security](https://www.humansecurity.com/) to prevent bots which does some stuff (~10k LOC of obfuscated Javascript) to build a signature of some opaque data that appears to be required for login. I don't see any easy way to work around this at the moment so unless someone discovers a workaround this is probably not fixable. If someone does discover a workaround though I'd be all in on rewriting Pydora with API compatibility on top of this new API.

**Comment by mcrute** [comment]
It looks pretty trivially easy to extract the auth tokens from the browser once you've authenticated with the website but they do expire (TTL seems to be about 12 hours) and I don't see any way to renew them without going back through the bot prevention logic, although it's possible I just haven't found the right call yet. Otherwise the API seems very straightforward so I might take a pass at an update client once I make it through some other high-priority items in other projects.

**Comment by zonywhoop** [comment]
I may be missing something but is there some reason the OAUTH2 user/browser or device flow cannot be used or implemented as selective options?

https://developer.pandora.com/docs/reference/graphql-api/authentication-using-oauth2/

**Comment by mcrute** [comment]
That would possibly work but each user would need a client ID and secret and Pandora isn’t issuing those right now. If I can get those credentials I could run a token broker on behalf of the Pydora project. Let me see if I can become an API partner.

**Comment by 89z** [comment]
You can use the Android client, I have an implementation here if it helps:

https://github.com/89z/mech/blob/master/pandora/pandora.go

**Comment by asterbini** [comment]
Yet Pithos works fine ... perhaps its Pandora authentication could inspire a solution?
https://github.com/pithos/pithos

**Comment by detroitgoose** [comment]
> TL;DR: Use the AIR client ID, this is probably not going to be fixable any time soon. ...In-fact the only credential set that seems to work for me anymore is the one from the AIR client. 

Same issue as thread topic, but AIR isn't working either. Is there something obviously wrong with my (Mopidy) config?

FWIW I have a Plus account, not Premium.

```
[pandora]
enabled = true
api_host = internal-tuner.pandora.com/services/json/
partner_encryption_key = 2%3WCL*JU$MP]4
partner_decryption_key = U#IO$RZPAB%VX2
partner_username = 'pandora one'
partner_password = TVCKIBGS9AO9TSYLNNFUML0743LH82D
partner_device = D01
username = xxxxx
password = xxxxx
```
Single or double quotes around `partner_username` produce the same error

```
pandora.errors.InvalidAuthToken: Invalid Auth Token
```

I really appreciate the work y'all put into this, so thanks for keeping on it.

**Comment by ksaye** [comment]
@asterbini, thanks for the suggestion.  I ported the python authentication to my simple 'music player' and while I loved Pydora, this works for now.

**Comment by theophileds** [comment]
@mcrute, I'm trying to manually set the `userAuthToken` in the APIClient, but I'm still encountering the exception: `pandora.errors.MissingAuthToken: Missing Auth Token`.

Do you have any practical examples of how to achieve that?

**Comment by ekimyar** [comment]
So the AIR client doesn't seem to work on pydora for me either (or the default client), but for some reason Pianobar *does* work-- I can login and play music with it on the same machine. Its sample config files seems to use the AIR client, at least in its sample config file. Unfortunately, the config file on my computers isn't in the default location, and I'm embarrassed to say that I can't find it to figure out what Pianobar is actually doing. I believe pithos uses pianobar for its backend, so it makes sense that pithos works.

**Comment by mcrute** [comment]
So far I've been operating under the assumption that this was a breaking server-side change but PianoBar still working would suggest that it's a bug here. Nothing jumps out at me in [PianoBar's implementation](https://github.com/PromyLOPh/pianobar/blob/master/src/libpiano/request.c) vs pydora's and nothing has substantially changed with the login code since 2019. I'll see if I can find some time to debug this and get a fix out this week.

**Comment by DeFrakt** [comment]
Any developments on this thread? Is using the new pandora oAuth to authorize client even an option?

**Comment by enzuru** [comment]
So when I dig into this bug, it actually appears to be failing with

> b'{"stat":"fail","code":1001,"message":"Invalid username and/or password"}'

I have tried to make the Pianobar and Pydora user login code exactly the same, but no cigar.

Around the time this occurred, users noticed this: https://community.pandora.com/t5/Desktop/Can-t-Log-In-on-Pithos/m-p/68237#M5322

Is it possible that Pandora whitelisted some unofficial clients, but not others? But we never identify ourselves differently, so how can they block Pydora and not Pandora?

**Comment by enzuru** [comment]
Aha, this fixed it:
```python
        headers = { "User-agent": "pianobar-2022.04.01" }
        result = self._http.post(url, data=data, params=params, headers=headers)
```
PR incoming.

**Comment by mcrute** [comment]
Thanks to the patch from @enzuru this should now be fixed in release 2.2.0 which is live now on PyPi. Everything seems to work but feel free to re-open this issue if the problem reoccurs.
